returnn/datasets/cached.py

```diff
@@ -444,9 +444,7 @@ def delete(self, nframes):
 
   @property
   def num_seqs(self):
-    if self._index_map:
-      return len(self._index_map)
-    return self._num_seqs
+    return len(self._index_map)
```

Comment:
Forgot to test with HDFDataset, after this `num_seqs=0` works also here.
`self. _index_map` is always set, so should be fine.

Reply:
Hmm, ok depends on what we want `Dataset.num_seqs` to be before `init_seq_order` is called, see the changed test case. `CachedDataset2` would raise `NotImplementedError`. Not sure if we want that here.

Reply:
You raise `NotImplementedError` when it is not implemented (or only in a derived class), and only then, not in any other case.

But I don't exactly understand anyway: If `self._index_map` is always set, this change here should not change anything?

Reply:
`__init__()` sets `self._index_map` to an empty range, which is falsy. So before an `init_seq_order()` call `CachedDataset.num_seq` is `self._num_seqs`. It would now be 0 after the change. Probably `None` would be the most logical value though?

> You raise `NotImplementedError` when it is not implemented

Yes, that would be nice. And that's why I don't want to use it. But what we have right now is `CachedDataset2` raising it depending on its state (whether `init_seq_order()` was called). And `ExternSprintDataset` even depending on the data loading speed...

Reply:
> But what we have right now is `CachedDataset2` raising it depending on its state

Yes, so where is the problem? Is there any contradiction? It means it is implemented depending on the state. I.e. it is not implemented for one particular case.

Reply:
So when `init_seq_order` was not called yet, as this is somewhat undefined anyway, you could also just leave the current code.